include/deal.II/base/tensor.h

```diff
@@ -909,7 +909,10 @@ inline
 typename Tensor<0,dim,Number>::real_type
 DEAL_II_CUDA_HOST_DEV Tensor<0,dim,Number>::norm_square () const
 {
+  // We cannot use Assert inside a CUDA kernel
+#ifndef DEAL_II_WITH_CUDA
```

Comment:
Oh. Out of curiosity, why not? And what should happen when one runs into an exceptional situation in a kernel?

Reply:
I only noted that we were guarding `Assert` for all host device functions in this file. These were introduced in #5737 by @rombur. Maybe, he can comment on this.

Reply:
To have the assert working in the kernel, we would need to rewrite our assert to use a special assert for cuda. This assert is not portable and does not work on mac. If we want to check that the kernel was run correctly we need to have some status variable that we need to check after the kernel was run. In practice, every time I have had a problem in a kernel, I got an error when I tried to get the result back on the host. So I know something went wrong...

Reply:
I don't think I have the knowledge to comment on what to do, but it is sad that (i) this means that we can't check errors on the GPU, and (ii) that the way you disable it here means that when deal.II is configured to use CUDA, we also disable the assertion on the CPU.

Long term it would be good to find a better solution where somewhere in the internals of the assertion mechanism we create the call to `std::abort()` on the CPU, and to whatever the magic equivalent is on the GPU.